lib/kernel-transport.js

```diff
@@ -0,0 +1,124 @@
+/* @flow */
+
+import { observable, action } from "mobx";
+
+import { log } from "./utils";
+
+export type ResultsCallback = (
+  message: any,
+  channel: "shell" | "iopub" | "stdin"
+) => void;
+
+export default class KernelTransport {
+  @observable executionState = "loading";
+  @observable inspector = { bundle: {} };
+
+  kernelSpec: Kernelspec;
+  grammar: atom$Grammar;
+  language: string;
+  displayName: string;
+  watchCallbacks: Array<Function> = [];
+
+  constructor(kernelSpec: Kernelspec, grammar: atom$Grammar) {
+    this.kernelSpec = kernelSpec;
+    this.grammar = grammar;
+
+    this.language = kernelSpec.language.toLowerCase();
+    this.displayName = kernelSpec.display_name;
+  }
+
+  @action
+  setExecutionState(state: string) {
+    this.executionState = state;
+  }
+
+  addWatchCallback(watchCallback: Function) {
+    this.watchCallbacks.push(watchCallback);
+  }
+
+  _callWatchCallbacks() {
+    this.watchCallbacks.forEach(watchCallback => watchCallback());
+  }
+
+  interrupt() {
+    throw new Error("KernelTransport: interrupt method not implemented");
+  }
+
+  shutdown() {
+    throw new Error("KernelTransport: shutdown method not implemented");
+  }
+
+  restart(onRestarted: ?Function) {
+    throw new Error("KernelTransport: restart method not implemented");
+  }
+
+  execute(code: string, callWatches: boolean, onResults: ResultsCallback) {
+    throw new Error("KernelTransport: execute method not implemented");
+  }
+
+  complete(code: string, onResults: ResultsCallback) {
+    throw new Error("KernelTransport: complete method not implemented");
+  }
+
+  inspect(code: string, cursorPos: number, onResults: ResultsCallback) {
+    throw new Error("KernelTransport: inspect method not implemented");
+  }
+
+  inputReply(input: string) {
+    throw new Error("KernelTransport: inputReply method not implemented");
+  }
+
+  isValidMessage(message: Message) {
```

Comment:
This check belongs to `ZMQKernel` (as far as I know, there is no need to check the validity of messages returned by `WSKernel`).

Reply:
Plugins can inject results messages that they generate. I thought it would be useful to validate these, so the check is needed by both `ZMQKernel` and `Kernel`.

That's why it's currently in this really awkward place of being defined for `KernelTransport`, which doesn't use it, and being public.

Reply:
The reason why we had these checks in `ZMQKernel` is because `ZMQKernel` can handle old kernels and kernels that aren't fully compliant with the Jupyter protocol. These checks helped us detect those kernels.

If we want to detect non-compliant plugins, then I'd remove the check from `ZMQKernel` and check the messages only once.

Reply:
I can remove the checks from `ZMQKernel` if you'd like. But won't there be an issue when users have a non-compliant kernel, and one of their plugins errors as a result? This would cause an issue to be reported to the plugin developer, which isn't necessarily right.

Reply:
Oh I see. I'll look into this: it's definitely the correct thing to do, just a question of how hard.